File: DsaUsingC++/bst.cpp
#include<bits/stdc++.h>
using namespace std;
class node{
public:
	int data;
	node * left=NULL;
	node * right =NULL;
};
class bst{
public:
	node * insert(node * root,int data);
	bool search(node * root,int data);
	node * create_node(int data);
	node * delete_node(node * root,int data);
	void inorder_display(node * root);
	void postorder_display(node * root);
	void preorder_display(node * root);
};
node * bst :: insert(node * root,int data){
	if(root==NULL){
		root=create_node(data);
		return root;
	}
	else if(root->data>=data){
		root->left=insert(root->left,data);
	}
	else{
		root->right=insert(root->right,data);
	}
	return root;
}
node * bst :: create_node(int data){
	node * temp=new node();
	temp->data=data;
	return temp;
}
void bst :: preorder_display(node * root){
	if(root==NULL){
		return;
	}
	cout<<root->data<<"->";
	inorder_display(root->left);
	inorder_display(root->right);

}
void bst :: inorder_display(node * root){
	if(root==NULL){
		return;
	}
	inorder_display(root->left);
	cout<<root->data<<"->";
	inorder_display(root->right);

}

void bst :: postorder_display(struct node* node)
{
     if (node == NULL)
        return;
     postorder_display(node->left);
     postorder_display(node->right);
     printf("%d->", node->data);
}
int main(){
	node *root=new node();
	bst b;
	b.insert(root,15);
	b.insert(root,10);
	b.insert(root,20);
	b.insert(root,25);
	b.insert(root,8);
	b.insert(root,12);
	b.inorder_display(root);
	cout<<"\n";
	b.postorder_display(root);
	cout<<"\n";
	b.preorder_display(root);
	cout<<"\n";
	b.inorder_display(root);
	cout<<"\n";
	return 0;
}

File: DsaUsingPython/tree/bst.py
class Node:
    def __init__(self,data):
        self.data=data
        self.left=None
        self.right=None

class bst:
    def create_node(self,data):
        temp=Node(data)
        return temp


    def insert_node(self,root,data):
        if root==None:
            root=self.create_node(data)
        elif(data<=root.data):
            if(root.left==None):
                root.left=self.create_node(data)
            else:
                self.insert_node(root.left,data)
        else:
            if(root.right==None):
                root.right=self.create_node(data)
            else:
                self.insert_node(root.right,data)


    def post_display(self,dnode):
        if(dnode==None):
            return
        else:
            self.post_display(dnode.left)
            self.post_display(dnode.right)
            print(dnode.data,end="->")



    def inorder_display(self,dnode):
        if(dnode==None):
            return
        else:
            self.inorder_display(dnode.left)
            print(dnode.data,end="->")
            self.inorder_display(dnode.right)


    def preorder_display(self,dnode):
        if dnode==None:
            return
        else:
            print(dnode.data,end="->")
            self.preorder_display(dnode.left)
            self.preorder_display(dnode.right)


    def bst_search(self,tree_root,data):
        if(tree_root is None):
            return(False)
        elif(tree_root.data == data):
            return (True)
        elif(data<=tree_root.data):
            return(self.bst_search(tree_root.left,data))
        else:
            return(self.bst_search(tree_root.right,data))


def main():
    tree=bst()
    tr=tree.create_node(0)
    tree.insert_node(tr,15)
    tree.insert_node(tr,10)
    tree.insert_node(tr,20)
    tree.insert_node(tr,25)
    tree.insert_node(tr,8)
    tree.insert_node(tr,12)

    print("postorder traversal of the above tree is ")
    tree.post_display(tr)
    print("\n")

    print("preorder traversal of the above tree is ")
    tree.preorder_display(tr)
    print("\n")

    print("inorder traversal of the above tree is ")
    tree.inorder_display(tr)
    print("\n")

    z=tree.bst_search(tr,12)
    if(z):
        print("number 12 found in the tree ")
    else:
        print("number 12 not found in the tree ")
#    print("\n")
    z=tree.bst_search(tr,145)
    if(z):
        print("number 145 found in the tree ")
    else:
        print("number 145 not found in the tree ")


if __name__ == '__main__':
    main()
